### ISSUE TITLE
Best practices for incremental migrations?

### ISSUE THREAD

**Issue by jjoelson** [simolus3/moor#2426]
Hi all! I know there have been some issues raised about potentially enhancing Drift to better support incremental migration (e.g. https://github.com/simolus3/drift/issues/1390), but I'm wondering if there are any current best practices or recommendations for doing this.

My initial approach to migrations involved a series of if-statements executed sequentially:
```
if (from < 2) {
  ...
}

if (from < 3) {
  ...
}

...
```

Each time the schema changes, I bump the schema version and add a new if-statement at the bottom to perform the migration to that version. This structure is very easy to reason about and very easy to implement; each new migration simply goes at the end after all of the previous migrations. 

However, this approach stops working once you have two migrations that add a new column to the same table using `alterTable`/`TableMigration`:
```
if (from < 2) {
  // Once `newColumn2` is added for schema version 3, this `alterTable` statement
  // starts failing with the following error:
  //     SqliteException(1): while executing, no such column: newColumn2, SQL logic error (code 1)
  await m.alterTable(
    TableMigration(
      someTable,
      newColumns: [
        someTable.newColumn1
      ],
      columnTransformer: {
        someTable.newColumn1: const Constant(''),
      },
    ),
  );
}

if (from < 3) {
  await m.alterTable(
    TableMigration(
      someTable,
      newColumns: [
        someTable.newColumn2
      ],
      columnTransformer: {
        someTable.newColumn2: const Constant(''),
      },
    ),
  );
}
```

My understanding is that Drift only knows the  _latest_ version of the schema, which leads to the error if a database is being updated from v1 to v3.

The solution that I came up with is to consolidate all table migrations into one statement, which looks something like this:
```
if (from < 3) {
  await m.alterTable(
    TableMigration(
      someTable,
      newColumns: [
        if (from < 2) someTable.newColumn1,
        if (from < 3) someTable.newColumn2,
      ],
      columnTransformer: {
        if (from < 2) someTable.newColumn1: const Constant(''),
        if (from < 3) someTable.newColumn2: const Constant(''),
      },
    ),
  );
}
```
However, this structure is much more difficult to reason about:
- Adding a new version migration involves going back and modifying existing migration code instead of tacking on new logic to the end
- If you have migrations involving multiple tables that have foreign key references to added columns, the logic to ensure statements are executed in the correct order gets out of hand quickly

Having tests for every migration scenario (v1 to v2, v1 to v3, etc.) is definitely a must and makes this manageable, but I can't help but feel there has to be a better way to handle this. Are there any best practices that can help?

**Comment by kuhnroyal** [comment]
If you reach the limits of the migrator API you can always switch to plain SQL. Or you can write a separate 1 to 3 migration that skips 2. But that gets tedious very fast. I stick with SQL only.

**Comment by jjoelson** [comment]
@kuhnroyal Thanks for sharing your experience! Sticking to SQL might be the way to go.

**Comment by simolus3** [comment]
A neat trick someone has shared a while back is to use the functionality for [schema test generation](https://drift.simonbinder.eu/docs/advanced-features/migrations/#generating-test-code) to generate snapshots of older versions of the database schema into `lib/`.

[This example](https://github.com/simolus3/drift/tree/develop/examples/migrations_example) in this repository makes use of that:

https://github.com/simolus3/drift/blob/f5ae3902846f7626d1bb2f5471f14f62319809dd/examples/migrations_example/lib/database.dart#L45-L51

This works really well if you have an exported snapshot of all your drift schemas, but it also generates a lot of code. Perhaps we could expand that into a tool that generates much simpler code into a single file by re-using table definitions that didn't change between schema versions.

**Comment by jjoelson** [comment]
@simolus3 That's a very interesting approach! Are there any concerns that backward-incompatible updates to Drift might invalidate the old schema files and make it impossible to generate up to date Dart code?

**Comment by simolus3** [comment]
I'm aware of the pain I'd cause by altering the format of those files. I think it's unlikely that there will ever be a need for a redesign since these files are essentially a JSON representation of the underlying sqlite3 schema with few drift-specific annotations. There's also a version tag in those files, so if something comes up, I can keep preserve backwards compatibility in the reader / `drift_dev` tools using those files.

**Comment by jjoelson** [comment]
@simolus3 My apologies for leaving this issue hanging! Thanks for for reassurance about preserving backward compatibility.

I think my takeaway from this discussion is that to avoid getting into trouble as schema versions pile up, one should either:
- Stick to custom SQL for all migrations
- Maintain snapshots of each schema version using the schema generation tool

Thank you!

**Comment by jjoelson** [comment]
Hi @simolus3! Circling back to this, I'm experimenting with the generated migrations as demonstrated in that example project, but I'm not seeing that `steps` command referenced in the README with the latest `drift_dev` version:
```
Could not find a subcommand named "steps" for "dart run drift_dev schema"
```
The help documentation only mentions `dump` and `generate` as available subcommands:
```
% dart run drift_dev schema -h
Inspect or manage the schema of a moor database

Usage: dart run drift_dev schema <subcommand> [arguments]
-h, --help    Print this usage information.

Available subcommands:
  dump       Export the entire table structure into a file
  generate   Generate Dart code to help verify schema migrations.

Run "dart run drift_dev help" to see global options.
```

Is this a command that is supposed to be available?

**Comment by simolus3** [comment]
It hasn't been released yet, I'll probably do that tomorrow. You can try it out by [overriding the dependencies](https://drift.simonbinder.eu/docs/internals/) to the latest development version though.

**Comment by simolus3** [comment]
The tool has been released with drift 2.10.

**Comment by jjoelson** [comment]
> The tool has been released with drift 2.10.

@simolus3 I finally got a chance to try this out and it's pretty slick! Thank you for all your hard work and continued active development on this package!

One question I have is whether the advice in the [tips](https://drift.simonbinder.eu/docs/advanced-features/migrations/#tips) section of the documentation about putting the migration in a transaction and then verifying foreign keys still applies when using the generated `stepByStep` migrations. I'm wondering because the API of the generated code seems intended to pass directly to `onUpgrade` like this:
```
onUpgrade: stepByStep(
  from1To2: ...,
),
```
But as far as I can tell the generated `stepByStep` code doesn't do anything internally with transactions or anything like that, so this is what I ended up writing:
```
  @override
  int get schemaVersion => 2;

  @override
  MigrationStrategy get migration {
    return MigrationStrategy(
      beforeOpen: (details) async {
        await customStatement('PRAGMA foreign_keys = ON');
      },
      onUpgrade: (m, from, to) async {
        await customStatement('PRAGMA foreign_keys = OFF');

        await transaction(
          () => _migrationSteps()(m, from, to),
        );

        if (kDebugMode) {
          final wrongForeignKeys =
              await customSelect('PRAGMA foreign_key_check').get();
          assert(wrongForeignKeys.isEmpty,
              '${wrongForeignKeys.map((e) => e.data)}');
        }
      },
    );
  }

  OnUpgrade _migrationSteps() => stepByStep(
        from1To2: (m, schema) async {
          await m.addColumn(schema.myTable, schema.myTable.myNewColumn);
        },
      );
```

Does that code look reasonable to you? The slight awkwardness of `_migrationSteps()(m, from, to)` has me wondering if this is the type of code you intended to be written using the generated `stepByStep` function.

**Comment by simolus3** [comment]
The code looks reasonable and correct to me, but I agree that it's kind of cumbersome.

895ff5276180d71e6d4cddcaaa750de25bd9bce0 is an attempt to allow more control over this, allowing you to customize the `onUpgrade` callback. You can see [this diff](https://github.com/simolus3/drift/commit/895ff5276180d71e6d4cddcaaa750de25bd9bce0#diff-14cad82d1a6db2f905a53a06555c19f5c595a543105e04a98009e0ea3e1bd5c6) for a comparison of the changes made to the example to enable this.
